packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -843,28 +873,61 @@ export const openTabFromCurrent = (
   };
 };
 
+type DuplicateTabAction = {
+  type: WorkspacesActions.DuplicateTab;
+  atIndex: number;
+};
+
+export const duplicateTab = (atIndex: number): DuplicateTabAction => {
+  return {
+    type: WorkspacesActions.DuplicateTab,
+    atIndex,
+  };
+};
+
+async function confirmClosingTabs() {
+  return await showConfirmation({
+    title: 'Are you sure you want to close the tab?',
+    description:
+      'The content of this tab has been modified. You will lose your changes if you close it.',
+    buttonText: 'Close tab',
+    variant: 'danger',
+    'data-testid': 'confirm-tab-close',
+  });
+}
+
 type CloseTabAction = { type: WorkspacesActions.CloseTab; atIndex: number };
 
 export const closeTab = (
   atIndex: number
 ): WorkspacesThunkAction<Promise<void>, CloseTabAction> => {
   return async (dispatch, getState) => {
     const tab = getState().tabs[atIndex];
-    if (!canCloseTab(tab)) {
-      const confirmClose = await showConfirmation({
-        title: 'Are you sure you want to close the tab?',
-        description:
-          'The content of this tab has been modified. You will lose your changes if you close it.',
-        buttonText: 'Close tab',
-        variant: 'danger',
-        'data-testid': 'confirm-tab-close',
-      });
-      if (!confirmClose) {
-        return;
+    if (canCloseTab(tab) || (await confirmClosingTabs())) {
+      dispatch({ type: WorkspacesActions.CloseTab, atIndex });
+      cleanupLocalAppRegistryForTab(tab?.id);
+    }
+  };
+};
+
+type CloseAllOtherTabsAction = {
+  type: WorkspacesActions.CloseAllOtherTabs;
+  atIndex: number;
+};
+
+export const closeAllOtherTabs = (
+  atIndex: number
+): WorkspacesThunkAction<Promise<void>, CloseAllOtherTabsAction> => {
+  return async (dispatch, getState) => {
+    const { tabs } = getState();
+    const otherTabs = tabs.filter((_, index) => index !== atIndex);
+    for (const tab of otherTabs) {
+      if (!canCloseTab(tab) && !(await confirmClosingTabs())) {
```

Comment:
Doesn't feel right that canceling close for one tab stops all the others from closing too (or the other way around, you say "yes" for one, then "no" for other, and all stay open), in vscode you'd be asked for evey one separately and they are closed based on your answers. Wouldn't that make more sense for our flow too?

Reply:
Also now that we have this case of confirmation being shown for multiple tabs, it would probably make sense to add something to this confirmation model to make identification easier. "Are you sure you want to close the tab?" reads fine if you actively clicked a close button on a certain tab, but not when you are getting this message as a side effect of another action with no context what exactly will be closed